src/cata_views.h

```diff
@@ -0,0 +1,86 @@
+#pragma once
+#ifndef CATA_SRC_CATA_VIEWS_H
+#define CATA_SRC_CATA_VIEWS_H
+
+#include <functional>
+
+namespace cata
+{
+namespace views
+{
+
+/*
+ * A lazily-evaluated transformation on an STL container for zero-copy iteration.
+ *
+ * Example:
+ *     std::vector<int> vec{ 1, 2, 3, 4, 5 };
+ *     auto square = []( const int &x ) -> int { return x * x; };
+ *     for( auto i : cata::views::transform<std::vector<int>, int>( vec, square ) ) {
+ *         // You get i = { 1, 4, 9, 16, 25 } in this ranged-based for loop,
+ *         // but avoids constructing a temporary std::vector<int> and copy over all elements
+ *     }
+ *
+ * You can even transform elements into a different type:
+ *     std::vector<intr> vec{ 1, 2, 3, 4, 5 };
+ *     auto stringify = []( const int &x ) -> std::string {
+ *         return string_format( _( "Square of %d is %d." ), x, x * x );
+ *     };
+ *     for( auto s : cata::views::transform<std::vector<int>, std::string>( vec, stringify ) ) {
+ *         you.add_msg_if_player( s );
+ *     }
+ *
+ * typename A is the type of the STL *container*, e.g. std::vector<mutable_overmap_placement_rule_piece>
+ * typename B is the type of *element* after transformation
+ */
+template<typename A, typename B>
+class transform
+{
+    private:
+        using fun_t = std::function<B( const typename A::value_type & )>;
+
+        fun_t fun;
```

Comment:
I see, so it's not a zero-allocation range, but it's definitely cheaper to have one member function than copying a whole set of values wholesale. I think part of the value (but also restriction) of `range-v3` is it's zero-allocation in many places.